## Manejo de datos

En este notebook se realizarán todas las acciones que correspondan a:  
- Transformación de tipo de variables
- Creación de nuevas variables
- Contabilizar valores nulos por columna
- Manejo de datos nulos
- Guardar el nuevo DataSet limpio

In [8]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
#warnings.filterwarnings('ignore')

In [9]:
df = pd.read_csv('C:/Users/valef/Desktop/DIPLOMADO UDD/python/hotel_booking.csv') #se carga el data set

In [11]:
df_cleaned = df.dropna(subset=['children', 'country']) #eliminación de datos nulos
#df_cleaned.info()
df_cleaned1 = df_cleaned #df de respaldo con las variables object en categorias

**Transformación de variables**

In [14]:
df['reservation_status_date']= pd.to_datetime(df['reservation_status_date']) #para transformar la variable a tipo fecha

**Creación de nuevas variables**

In [22]:
#creación de la variable 'date_checkin' que indica la echa que se registran los huespedes
mes_map = {
    'January': 1,
    'February': 2,
    'March': 3,
    'April': 4,
    'May': 5,
    'June': 6,
    'July': 7,
    'August': 8,
    'September': 9,
    'October': 10,
    'November': 11,
    'December': 12
}
df['mes_num'] = df['arrival_date_month'].map(mes_map)

df['year'] = df['arrival_date_year']
df['month'] = df['mes_num']
df['day'] = df['arrival_date_day_of_month']


df['date_checkin'] = pd.to_datetime(df[['year', 'month', 'day']])


In [26]:
#para calcular la cantidad de dias de la estadia restando las fechas de salida y entrada
df['dias_estadia'] = (df['reservation_status_date'] - df['date_checkin']).dt.days

In [144]:
#otra forma de calcular la cantidad de dias de la estadia sumando stays_in_weekend_nights y stays_in_week_nights
df['dias_estadia_1'] = (df['stays_in_weekend_nights'] + df['stays_in_week_nights'])

In [28]:
#nuevas variables respecto a la cantidad de huespedes "niños" y "niños + adultos"
df['kids'] = (df['babies'] + df['children'])
df['num_huespedes'] = (df['kids'] + df['adults'])

In [30]:
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  phone-number  \
0                        0                     0       2  ...  669-792-1661   
1                        0                     0       2  ...  858-637-6955   
2                        0                     1       1  ...  652-885-2745   
3                        0                     1       1  ...  364-656-8427   
4                        0                     2       2  ...  713-226-5883   

        credit_card mes_num  year month day  date_checkin  dias_estadia  kids  \
0  ************4322       7  2015     7   1    2015-07-01             0   0.0   
1  ************9157       7  2015     7   1    2015-07-01             0   0.0   
2  ************3734       7  2015     7   1    2015-07-01             1   0.0   
3  ************5677       7  2015     7   1    2015-07-01             1   0.0   
4  ************5498       7  2015     7   1    2015-07-01             2   0.0   

  num_huespedes  
0           2.0  
1           2.0  
2           1.0  
3           1.0  
4           2.0  

[5 rows x 44 columns]

In [76]:
df.info() # verifico la creación de variables

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 44 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   hotel                           119390 non-null  category      
 1   is_canceled                     119390 non-null  int64         
 2   lead_time                       119390 non-null  int64         
 3   arrival_date_year               119390 non-null  int64         
 4   arrival_date_month              119390 non-null  category      
 5   arrival_date_week_number        119390 non-null  int64         
 6   arrival_date_day_of_month       119390 non-null  int64         
 7   stays_in_weekend_nights         119390 non-null  int64         
 8   stays_in_week_nights            119390 non-null  int64         
 9   adults                          119390 non-null  int64         
 10  children                        119386 non-null  float64

In [62]:
print(df.isnull().sum())

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

## Datos Nulos.

Una vez creadas las nuevas variables, existen 6 columnas que presentan datos nulos o vacios:  
- **children:** 4 filas
- **country:** 488 filas
- **agent:** 16.340 filas
- **company:** 112.593 filas
- **kids:** 4 filas, como consecuencia de 'children'
- **num_huespedes:** 4 filas, como consecuencia de 'children'

En el caso de agent y company no realizaremos ninguna acción ya que no son relevantes para nuestro análisis. Sin embargo, en el caso de "children" y "country" eliminaremos las filas que contengan valores nulos, esta decisión de debe a que solamente estaremos eliminando 492 filas, equivalentes al 0.41% de las filas, siendo considerado esta cifra marginal respecto del total. Por otro lado, al eliminar las filas de 'children' se eliminaran los nulos de 'kids' y 'num_huespedes'.

In [64]:
df_cleaned = df.dropna(subset=['children', 'country']) #eliminación de datos nulos de las variables "children" y "country"

In [66]:
print(df_cleaned.isnull().sum()) #verifico nulos. Solo dejamos lode agent y company

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               0
babies                                 0
meal                                   0
country                                0
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16004
company         

**Una vez realizados los cambios (solucionar NULL e incorporar nuevas variables) se guarda el nuevo DataSet para registro.**

In [68]:
df_cleaned.to_csv('data_cleaned.csv', index=False) #se guarda

In [70]:
df_clean = pd.read_csv('C:/Users/valef/Desktop/DIPLOMADO UDD/python/ProyectoPython/NOTEBOOK/data_cleaned.csv') #se carga el DataSet

In [72]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118898 entries, 0 to 118897
Data columns (total 44 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           118898 non-null  object 
 1   is_canceled                     118898 non-null  int64  
 2   lead_time                       118898 non-null  int64  
 3   arrival_date_year               118898 non-null  int64  
 4   arrival_date_month              118898 non-null  object 
 5   arrival_date_week_number        118898 non-null  int64  
 6   arrival_date_day_of_month       118898 non-null  int64  
 7   stays_in_weekend_nights         118898 non-null  int64  
 8   stays_in_week_nights            118898 non-null  int64  
 9   adults                          118898 non-null  int64  
 10  children                        118898 non-null  float64
 11  babies                          118898 non-null  int64  
 12  meal            

In [52]:
df_clean.head

<bound method NDFrame.head of                hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
118893    City Hotel            0         23               2017   
118894    City Hotel            0        102               2017   
118895    City Hotel            0         34               2017   
118896    City Hotel            0        109               2017   
118897    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
